src/tests/nativeaot/CustomMain/CustomMainNative.cpp

```diff
@@ -0,0 +1,22 @@
+
+#include <stdio.h>
+#include <stdint.h>
+
+#if defined(_WIN32)
+extern "C" int __managed__Main(int argc, wchar_t* argv[]);
+#else
+extern "C" int __managed__Main(int argc, char* argv[]);
+#endif
+
+extern "C" void SetExitCodeInManagedSide(int32_t exitCode);
+
+#if defined(_WIN32)
+int __cdecl wmain(int argc, wchar_t* argv[])
+#else
+int main(int argc, char* argv[])
+#endif
+{
+    puts("hello from native code");
+    SetExitCodeInManagedSide(100);
```

Comment:
Maybe I did not understand the intention correctly, but shouldn't `NativeAOT_StaticInitialization` be called prior to any managed side code.

Reply:
See the changes to the MSBuild files. They force the `NativeAOT_StaticInitialization` symbol to be defined. This causes the constructor of the `InitializeRuntimePointerHelper` class to run before the native main function. This registers a callback with the native AOT runtime. On the first reverse P/Invoke into managed code, the NativeAOT runtime will run the callback to initialize the runtime.

The `__managed_Main` function is also a reverse P/Invoke, so initialization can happen when it is called too.

Reply:
Ah my bad, I missed that part. Thanks for pointing it out.